### PR TITLE
[Bugfix] Offline mode fix

### PR DESCRIPTION
**Pull request by joerunde**
This PR fixes a bug that I believe was introduced by the mistral model format support in https://github.com/vllm-project/vllm/pull/8168

The transformers_utils/config.py file now uses an unsafe `file_exists` check that will immediately fail in offline mode. This takes down all of our deployments where we set `HF_HUB_OFFLINE=1` to ensure that we don't pull model files from the hub.

The changes here first add a check to look for the model revision in local cache, so that no call out to the hub is needed when the model is already cached (⚡ ⚡ performance improvement). Then, these changes guard against the `OfflineModeIsEnabled` exception so that we can properly loop through the different model types without crashing early.

I've added a test case that currently fails on main, which tries to load a model in offline mode. Getting the test setup correct is a bit janky, as far as I can tell the only way to enable offline mode on the fly is to recursively reload both `transformers` and `huggingface_hub`, since some objects are initialized at import-time that determine whether or not they operate in offline mode.

FIX (I didn't open a bug 😞 ) 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/transformers_utils/config.py
Comment (joerunde) [inline vllm/transformers_utils/config.py:73]: FYI I _think_ this was a bug before, since for mistral models we'd want to be looking for the mistral config name on the hub

**Comment by joerunde** [comment]
Ah dang, my attempt to re-import things after the offline test is failing 🤔

**Review by mgoin — APPROVED** [review]

**Review by mgoin — COMMENTED** [review]
I think it might be easiest to make a completely new test entry that checks multiple things with HF_HUB_OFFLINE set. This would allow you to get past the tricky patching

**Comment by joerunde** [comment]
> I think it might be easiest to make a completely new test entry that checks multiple things with HF_HUB_OFFLINE set. This would allow you to get past the tricky patching

Yeah, separating out offline mode tests to run by themselves sounds way better. I think there's still the issue of making sure that the cached files exist ahead of time though, I imagine that CI step would have to invoke a download script first and then invoke pytest with offline mode enabled 🤔. I was hoping to find a more portable solution so everybody can just `pytest` this on their machines...

I'll hack away for a bit

**Review by mgoin — CHANGES_REQUESTED** [review]
I understand your thoughts, appreciate it and requesting changes for now to signify it isn't ready

**Comment by joerunde** [comment]
> I understand your thoughts, appreciate it and requesting changes for now to signify it isn't ready

Alright, hopefully this works- I've separated the test out into it's own module under tests/entrypoints and added a separate pytest invocation to the test pipeline for it. I've still kept the import hackery in so that the test is nicely encapsulated, and also fixed the last problem I was seeing with the imports and made the failure a bit more robust. (That's probably moot because I doubt many other tests will be added here, but I didn't want to leave it in an unhealthy state)

🤞  it all passes

**Review by njhill — APPROVED** [review]
Thanks @joerunde, LGTM! I will let @mgoin check again before we merge it though

**Comment by joerunde** [comment]
```
docker: failed to register layer: write /opt/conda/envs/py_3.9/lib/python3.9/site-packages/transformers/models/maskformer/modeling_maskformer.py: no space left on device.
```

uhhhhh.... :(

I think I'll need somebody with write access to retry those failures?

**Comment by joerunde** [comment]
Looks like all the failures now are due to no space left on device, and I see builds on main failing with backoffs on docker pulls :(

I won't trigger a rebuild here- don't want to make these CI stability problems worse

**Comment by njhill** [comment]
Test failures are unrelated.

**Review by mgoin — APPROVED** [review]
LGTM, thanks for the test!

**Pull request merged**
